app/lib/discourse_automation/scripts/add_user_to_group_through_custom_field.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+DiscourseAutomation::Scriptable::ADD_USER_TO_GROUP_THROUGH_CUSTOM_FIELD = 'add_user_to_group_through_custom_field'
+
+# This script takes the name of a User Custom Field containing a group name.
+# On each run, it ensures that each user belongs to the group name given by that UCF.
+#
+# In other words, it designates a certain User Custom Field to act as
+# a "pointer" to a group that the user should belong to, and adds users as needed.
+
+DiscourseAutomation::Scriptable.add(DiscourseAutomation::Scriptable::ADD_USER_TO_GROUP_THROUGH_CUSTOM_FIELD) do
+  field :custom_field_name, component: :text, required: true
+
+  version 1
+
+  triggerables %i[recurring]
+
+  script do |trigger, fields|
+    custom_field_name = fields.dig('custom_field_name', 'value')
+
+    query = DB.query(<<-SQL, custom_field_name)
+      SELECT u.id as user_id, g.id as group_id
+      FROM users u
+      JOIN user_custom_fields ucf
+        ON u.id = ucf.user_id
+        AND ucf.name = ?
+      JOIN groups g
+        on g.name = ucf.value
+      FULL OUTER JOIN group_users gu
+        ON gu.user_id = u.id
+        AND gu.group_id = (select id from groups where name = ucf.value)
```

Comment:
May I know why we can't replace it with `gu.group_id = g.id`? We already filtered it by group name in the `groups` table's join query, right?

If it's making problems in the outer join query then can't prevent it by the `AND gu.user_id IS NOT NULL` condition?

Reply:
You're totally right - wow, it's painfully obvious now!